expression/integration_test.go

```diff
@@ -8027,3 +8027,22 @@ func TestIssue50850(t *testing.T) {
 		testkit.Rows("01", "31", "34", "5D", "65", "A5", "A6", "B1", "D5", "FF"))
 	tk.MustExec("drop table if exists t3;")
 }
+
+func TestIssue51765(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+
+	tk.MustExec("use test")
+	tk.MustExec("create table t (id varbinary(16))")
+	tk.MustExec("create table t1(id char(16) charset utf8mb4 collate utf8mb4_general_ci)")
+	tk.MustExec("insert into t values ()")
+	tk.MustExec(`insert into t1 values ("Hello World")`)
+
+	tk.MustQuery("select collation(ifnull(concat(NULL), '~'))").Check(testkit.Rows("utf8mb4_bin"))
```

Comment:
The result is different from the original PR in integration test, because the connection collation of Go MySQL driver is `utf8mb4_general_ci`, but here it's the default value `utf8mb4_bin`. (For mysql cli client, it's `utf8mb4_0900_ai_ci`).